include/sparrow/layout.hpp

```diff
@@ -0,0 +1,251 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or mplied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <concepts>
+#include <cstdint>
+
+#include "sparrow/array_data.hpp"
+#include "sparrow/iterator.hpp"
+#include "sparrow/buffer.hpp"
+
+namespace sparrow
+{
+    template <class T, bool is_const>
+    class primitive_layout_iterator
+        : public iterator_base
+        <
+            primitive_layout_iterator<T, is_const>,
+            mpl::constify_t<T, is_const>,
+            std::contiguous_iterator_tag
+        >
+    {
+    public:
+
+        using self_type = primitive_layout_iterator<T, is_const>;
+        using base_type = iterator_base
+        <
+            self_type,
+            mpl::constify_t<T, is_const>,
+            std::contiguous_iterator_tag
+        >;
+        using pointer = typename base_type::pointer;
+
+        // Required so that std::ranges::end(p) is
+        // valid when p is a primitive_layout.
+        primitive_layout_iterator() = default;
+        explicit primitive_layout_iterator(pointer p);
+
+    private:
+
+        // TODO: use reference proxies with an API similar to `std::optional` instead.
+        using reference = typename base_type::reference;
+        using difference_type = typename base_type::difference_type;
+        using size_type = std::size_t;
+
+        reference dereference() const;
+        void increment();
+        void decrement();
+        void advance(difference_type n);
+        difference_type distance_to(const self_type& rhs) const;
+        bool equal(const self_type& rhs) const;
+        bool less_than(const self_type& rhs) const;
+
+        // We only use the first buffer and the bitmap.
+        pointer m_pointer = nullptr;
+        size_type m_index = 0;
```

Comment:
If I understand correctly, `m_pointer`  is the currently pointed element, but then we have an index instead of just changing `m_pointer`. Why this choice? Why not just iterate the pointer?

Reply:
Indeed, before an alternative approach was used, yet `m_index` can be removed.